### PR TITLE
feat(ui): inline virtual text

### PR DESCRIPTION
**Pull request by bfredl**
Note: This is just the first patch. there are many vim patches to go through which fixes various edge cases and bugs.

### CODE REVIEW

File: src/nvim/plines.c [hunk 1]
```diff
@@ -296,8 +296,16 @@ void init_chartabsize_arg(chartabsize_T *cts, win_T *wp, linenr_T lnum FUNC_ATTR
   cts->cts_line = line;
   cts->cts_ptr = ptr;
   cts->cts_cur_text_width = 0;
-  // TODO(bfredl): actually lookup inline virtual text here
```
Comment (zeertzjq) [inline src/nvim/plines.c:299]: The doxygen comment above this function may also need change to stop mentioning "text properties", and `FUNC_ATTR_UNUSED` can be removed now.

**Comment by Jesse-Bakker** [comment]
On difference between this and the included vim patches seems to be redrawing the cursor after inserting inline virtual text before it.

A failing test case:
```lua
    it('adjusts cursor location correctly when inserting around inline virtual text', function()
        insert('12345678')
        feed '$'
        meths.buf_set_extmark(0, ns, 0, 4,
            { virt_text = { { ' virtual text ', 'Special' } }, virt_text_pos = 'inline' })

        screen:expect { grid = [[
      1234{28: virtual text }567^8                            |
      {1:~                                                 }|
      {1:~                                                 }|
      {1:~                                                 }|
      {1:~                                                 }|
      {1:~                                                 }|
      {1:~                                                 }|
      {1:~                                                 }|
      {1:~                                                 }|
      {1:~                                                 }|
      {1:~                                                 }|
      {1:~                                                 }|
      {1:~                                                 }|
      {1:~                                                 }|
                                                        |
      ]]
        }
end)
```

Instead of drawing the cursor at the end of the line, it is drawn in the middle of the virtual text until it is updated by moving the cursor left or entering insert mode.

**Comment by xiantang** [comment]
need to consider the case of `gj` and `gk` which are moved between the display line, I patched this pr, but actually, it works strangely. I supposed need to consider the virtual text in display line.

**Comment by Jesse-Bakker** [comment]
A dirty fix:

```diff
diff --git a/src/nvim/decoration.c b/src/nvim/decoration.c
index 037eb9f0d..96895bc1b 100644
--- a/src/nvim/decoration.c
+++ b/src/nvim/decoration.c
@@ -78,6 +78,13 @@ void decor_redraw(buf_T *buf, int row1, int row2, Decoration *decor)
       redraw_buf_range_later(buf, row1 + 1, row2 + 1);
     }
   }
+  if (decor && decor_virt_pos(*decor) && decor->virt_text_pos == kVTInline) {
+        FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
+            if (wp->w_cursor.lnum == row1 + 1) {
+                wp->w_valid &= ~(VALID_VIRTCOL | VALID_WROW | VALID_WCOL);
+            }
+        }
+    }

   if (decor && decor_virt_pos(*decor)) {
     redraw_buf_line_later(buf, row1 + 1, false);
```

**Comment by rcarriga** [comment]
Lots of discussion here so bit of an info dump, apologies.

> We could even just return the cancel function directly but returning a table leaves it more open to extending in future.

Not sure what I was thinking here when I said this. We can just return a task ID and provide module level functions for the cancelling etc, which would probably fit much more with the rest of core's APIs.

```lua
local task_id = async.run(function() end)
async.cancel(task_id)
```

---


> Hmm, I would try and tweak this to something like:

I'm not sure overloading a `pcall` function would make the most sense. There's a a couple of points that stick out to me:
Passing a task to pcall is a bit deceptive because the task is already running. It would also add a fair bit of boilerplate if making several calls and when adding arguments

```lua
local success, result1 = async.pcall(async.run(async_func_1))

local success, result2 = async.pcall(async.run(function()
  async_func_2(arg1)
end))

local function proxying_func(...)
  local args = {...}
  local success, result = async.pcall(async.run(function()
    async_func_3(unpack(args))
  end))
end
```

If we just pass async functions directly, which would be common in async contexts then we have to decide how to handle arguments

We could allow passing arguments
```lua
local success, result1 = async.pcall(async_func_1)

local success, result2 = async.pcall(async_func_2, arg1)

local function proxying_func(...)
  local success, result = async.pcall(async_func_3, ...)
end
```

Which is great but if we want to allow calling it in a non async context then we'd either have to require the number of arguments to be provided or provide the callback before the arguments

```lua
-- Provide number of arguments
async.pcall(async_func_1, 0, function(success, result))
    ...
end)

async.pcall(async_func_2, 1, arg1, function(success ,result)
    ...
end)

-- Provide callback before args
async.pcall(async_func_1, function(success, result))
    ...
end)

async.pcall(async_func_2, function(success ,result)
    ...
end, arg1)
```

Both of which require changing the type signature in IMO a clunky way that would not play well with the LSP. Also might lead to some surprising behaviour when calling from an async context but trying to provide a callback.

Sorry to be so verbose on this, just wanted to illustrate my thoughts. I think keeping `pcall` as an async only function makes the API much simpler.

---

> The resolution was that we should just be more strict on naming conventions and/or annotate the code better with things like -- @async

Will also add that the Lua LSP has an inline hint for `await` when calling functions marked as `async` so users can use that to see awaits easily. When https://github.com/neovim/neovim/pull/20130 is merged, it'll be an even better experience as they can be inline.

**Comment by tom-anders** [comment]
@bfredl I'd like to help with this PR if possible, is there a list of vim patches that we need to port over for this? Probably just `git log --grep="virtual text"` should be a good start?

**Comment by glepnir** [comment]
as far as i know in vim it's `textprop`

**Comment by luukvbaal** [comment]
Grepping the git log is a start but not every commit message for a relevant patch is guaranteed to contain those patterns. I think there is no other way but to actually skim through the entire log and make a judgement call of whether it could be relevant before inspecting the commit.

At least that's what I did for porting over the 'smoothscroll' [patches](https://github.com/neovim/neovim/pull/23320/commits), there are several that don't contain the word 'smoothscroll'.

**Comment by tom-anders** [comment]
I think I'll start with trying to fix https://github.com/neovim/neovim/issues/20524 - There's a corresponding functional test case in this PR that currently fails (`fold attached virtual lines are drawn correctly #21837`)

**Comment by SleepySwords** [comment]
If you want to take a look, I have a [branch](https://github.com/SleepySwords/neovim/tree/inline2) that I believe contains some of the main patches (I probably did miss quite a bit, this was only searching `virtual text` on GitHub commits, there are also some patches that I still am not sure if they need to be implemented). It is functional enough that I sometimes use this for inlay hints while working on projects. Not too sure if the code quality is up to standard. There are also still a lot of edge cases that need to be fixed.

Some stuff that is still broken/implemented:
- [ ] RHS does not seem to work
- [ ] If a virtual text is slightly, but not fully overlapping with `showbreak` enabled, the alignment of the first character is off. (happens in vim)
- [ ] Emphermal does not work (Not sure what the solution might be?)
- [ ] If `list` is enabled with an extend char and the virtual text is perfectly outside, the extend character is not displayed (also happens in vim)
- [ ] Deleting characters causes weird drawing (https://github.com/vim/vim/issues/12244)
- [ ] Probably some other extmark options (is something like `hl_mode` relevant?)

Probably even more that are not found yet, especially with multiple virtual texts.
Just as a note: while fixing failing tests, I messed up the initial commit, which is then fixed later (https://github.com/neovim/neovim/commit/98ac518f74f6d8352a266acd448672b57ae1a9b2)

Hope this helps!

**Comment by bfredl** [comment]
@SleepySwords great work! 

Not everything needs to work in a MVP PR. We should start with completely disallowing inline virtual text in ephemeral marks (and then consider limited functionality as a follow up).

Not all edge cases need to be "right" immediately. What's important is than we do have tests for the ones that work (by having specific code in patches, etc). Then we can get something merged and iterate over the 0.10 cycle.

**Comment by tom-anders** [comment]
> If you want to take a look, I have a [branch](https://github.com/SleepySwords/neovim/tree/inline2) that hopefully contains most of the main patches (I probably did miss quite a bit, this was only searching `virtual text` on GitHub commits, where some patches were skipped since I still was not sure if they need to be implemented or were relevant). It is functional enough that I sometimes use this for inlay hints while working on projects. Not too sure if the code quality is up to standard, and probably does contain dodgy fixes. There are also still a lot of edge cases that need to be fixed (RHS doesn't work, ephemeral doesn't work, weird edge cases regarding `showbreak` and `list with extends`, etc).
> 
> Just as a note: while fixing failing tests, I messed up the initial commit, which is then fixed later [98ac518](https://github.com/neovim/neovim/commit/98ac518f74f6d8352a266acd448672b57ae1a9b2) (Kinda dodgy fix)
> 
> Hope this helps!

Thanks! Just for my understanding: what does RHS refer to in this context?

**Comment by SleepySwords** [comment]
> Thanks! Just for my understanding: what does RHS refer to in this context?

Sorry I meant to say `rightleft`. I think I have misunderstood how `rightleft` is supposed to work and it seems fine? Is inserting one character after the cursor expected behaviour? It looked a bit odd with the vertical bar cursor.

File: src/nvim/plines.h [hunk 2]
```diff
@@ -11,9 +11,12 @@ typedef struct {
...
   char *cts_line;    // start of the line
   char *cts_ptr;     // current position in line
+  int cts_row;
 
   bool cts_has_virt_text;  // true if if a property inserts text
-  int cts_cur_text_width;     // width of current inserted text
+  int cts_cur_text_width_left;     // width of virtual text left of cursor
+  int cts_cur_text_width_right;     // width of virtual text right of cursor
+  MarkTreeIter cts_iter[1];
   // TODO(bfredl): iterator in to the marktree for scanning virt text
```
Comment (zeertzjq) [inline src/nvim/plines.h:20]:
```suggestion
```

**Comment by bfredl** [comment]
Merged, thanks @SleepySwords , @tom-anders and everyone else who have contributed with code and testing!

As always with a master feature, API and feature details is subject to change until next release (0.10) feature freeze. With this in mind, early testing and bug reports from early plugin adopters would be very welcome.

**Comment by nyngwang** [comment]
Oh my god, thank you so much!!!!

**Comment by B4mbus** [comment]
Finally! Now I can die in peace!

**Comment by clason** [comment]
So can this PR.

**Pull request merged**